src/zkp/ecdsa_adaptor.rs

```diff
@@ -0,0 +1,206 @@
+}
+
+impl EcdsaAdaptorSignature {
+    /// Creates an adaptor signature along with a proof to verify the adaptor signature.
+    pub fn adaptor_encrypt<C: Signing>(
+        secp: &Secp256k1<C>,
+        msg: &Message,
+        sk: &SecretKey,
+        enckey: &PublicKey,
+    ) -> EcdsaAdaptorSignature {
+        let mut adaptor_sig = ffi::EcdsaAdaptorSignature::new();
+
+        unsafe {
+            assert_eq!(
+                1,
+                ffi::secp256k1_ecdsa_adaptor_encrypt(
+                    *secp.ctx(),
+                    &mut adaptor_sig,
+                    sk.as_c_ptr(),
+                    enckey.as_c_ptr(),
+                    msg.as_c_ptr(),
+                    None,
+                    ptr::null_mut(),
+                )
+            );
+        }
+
+        EcdsaAdaptorSignature(adaptor_sig)
+    }
+
+    /// Creates an ECDSA signature from an adaptor signature and an adaptor secret.
+    pub fn adaptor_decrypt(&self, decryption_key: &SecretKey) -> Signature {
+        unsafe {
+            let mut signature = ffi::Signature::new();
+            assert_eq!(
+                1,
+                ffi::secp256k1_ecdsa_adaptor_decrypt(
+                    ffi::secp256k1_context_no_precomp,
+                    &mut signature,
+                    decryption_key.as_c_ptr(),
+                    self.as_c_ptr(),
+                )
+            );
+            Signature::from(signature)
+        }
+    }
+
+    /// Extracts the adaptor secret from the complete signature and the adaptor signature.
+    pub fn adaptor_recover<C: Signing>(
+        &self,
+        secp: &Secp256k1<C>,
+        sig: &Signature,
+        encryption_key: &PublicKey,
+    ) -> Result<SecretKey, Error> {
+        let mut data: [u8; constants::SECRET_KEY_SIZE] = [0; constants::SECRET_KEY_SIZE];
+
+        unsafe {
+            assert_eq!(
+                1,
+                ffi::secp256k1_ecdsa_adaptor_recover(
+                    *secp.ctx(),
+                    data.as_mut_c_ptr(),
+                    sig.as_c_ptr(),
+                    self.as_c_ptr(),
+                    encryption_key.as_c_ptr()
+                )
+            );
+        }
+
+        Ok(SecretKey::from_slice(&data)?)
+    }
+
+    /// Verifies that the adaptor secret can be extracted from the adaptor signature and the completed ECDSA signature.
+    pub fn adaptor_verify<C: Verification>(
+        &self,
+        secp: &Secp256k1<C>,
+        msg: &Message,
+        pubkey: &PublicKey,
+        encryption_key: &PublicKey,
+    ) -> Result<(), Error> {
+        unsafe {
+            let res = ffi::secp256k1_ecdsa_adaptor_verify(
+                *secp.ctx(),
+                self.as_c_ptr(),
+                pubkey.as_c_ptr(),
+                msg.as_c_ptr(),
+                encryption_key.as_c_ptr(),
+            );
+            return if res == 1 {
+                Ok(())
+            } else {
+                Err(Error::Upstream(UpstreamError::IncorrectSignature))
```

Comment:
If we introduce another variant for handling the cases of passing in a bad secret-key or signature, we might also want to use that because strictly speaking the adaptor sig is not "malformed" it just doesn't fit together.

Open for suggestions here in how to best handle this. I do think that `Error::Upstream` is not a good choice here because that variant only exists so that we can call into upstream functionality and properly return those errors. We shouldn't be constructing those ourselves because it is misleading as they don't actually originate upstream then.

Reply:
I see your point, thanks for the explanation. I changed to use `InvalidEcdsaAdaptorSignature` for all errors related to this module for now. I'm not sure it's the best, but at the same time I'm also not sure it would be better to have `FailedToRecover`, `FailedToDecrypt`... especially thinking that we can't really tell what the issue is apart that the native code returned an error. If you prefer the multiple error approach I don't mind updating it though.